core/lib/types/src/commitment/mod.rs

```diff
@@ -0,0 +1,666 @@
+    let mut input = vec![0_u8; final_len];
+    input[0..4].copy_from_slice(&(values.len() as u32).to_be_bytes());
+
+    let chunks = input[4..].chunks_mut(I::SERIALIZED_SIZE);
+    for (value, chunk) in values.iter().zip(chunks) {
+        value.serialize_commitment(chunk);
+    }
+    input
+}
+
+/// Serialize elements for commitment. The result consists of packed serialized elements.
+pub fn serialize_commitments<I: SerializeCommitment>(values: &[I]) -> Vec<u8> {
+    let final_len = values.len() * I::SERIALIZED_SIZE;
+    let mut input = vec![0_u8; final_len];
+
+    let chunks = input.chunks_mut(I::SERIALIZED_SIZE);
+    for (value, chunk) in values.iter().zip(chunks) {
+        value.serialize_commitment(chunk);
+    }
+    input
+}
+
+/// Precalculated data for the L1 batch that was used in commitment and L1 transaction.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct L1BatchMetadata {
+    pub root_hash: H256,
+    pub rollup_last_leaf_index: u64,
+    pub merkle_root_hash: H256,
+    pub initial_writes_compressed: Option<Vec<u8>>,
+    pub repeated_writes_compressed: Option<Vec<u8>>,
+    pub commitment: H256,
+    pub l2_l1_merkle_root: H256,
+    pub block_meta_params: L1BatchMetaParameters,
+    pub aux_data_hash: H256,
+    pub meta_parameters_hash: H256,
+    pub pass_through_data_hash: H256,
+    /// The commitment to the final events queue state after the batch is committed.
+    /// Practically, it is a commitment to all events that happened on L2 during the batch execution.
+    pub events_queue_commitment: Option<H256>,
+    /// The commitment to the initial heap content of the bootloader. Practically it serves as a
+    /// commitment to the transactions in the batch.
+    pub bootloader_initial_content_commitment: Option<H256>,
+    pub state_diffs_compressed: Vec<u8>,
+}
+
+impl L1BatchMetadata {
+    pub fn tree_data(&self) -> L1BatchTreeData {
+        L1BatchTreeData {
+            hash: self.root_hash,
+            rollup_last_leaf_index: self.rollup_last_leaf_index,
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct L1BatchWithMetadata {
+    pub header: L1BatchHeader,
+    pub metadata: L1BatchMetadata,
+    pub raw_published_factory_deps: Vec<Vec<u8>>,
+}
+
+impl L1BatchWithMetadata {
+    pub fn new(
+        header: L1BatchHeader,
+        metadata: L1BatchMetadata,
+        unsorted_factory_deps: HashMap<H256, Vec<u8>>,
+        raw_published_bytecode_hashes: &[H256],
+    ) -> Self {
+        Self {
+            raw_published_factory_deps: Self::factory_deps_in_appearance_order(
+                &header,
+                unsorted_factory_deps,
+                raw_published_bytecode_hashes,
+            ),
+            header,
+            metadata,
+        }
+    }
+
+    /// Iterates over factory deps in the order in which they appeared in this L1 batch.
+    pub fn factory_deps_in_appearance_order(
```

Comment:
Nit: This function can be private (unless it's going to be use somewhere in the future).